pkg/router/template/template_helper_test.go

```diff
@@ -1067,3 +1067,105 @@ func TestParseIPList(t *testing.T) {
 		})
 	}
 }
+
+func TestMaxTimeoutFirstMatchedAndClipped(t *testing.T) {
+	testCases := []struct {
+		name       string
+		state      map[ServiceAliasConfigKey]ServiceAliasConfig
+		annotation string
+		pattern    string
+		values     []string
+		expected   string
+	}{
+		{
+			name: "Route timeout is maximum",
+			state: map[ServiceAliasConfigKey]ServiceAliasConfig{
+				"test:route1": {
+					Annotations: map[string]string{
+						"haproxy.router.openshift.io/timeout": "5m",
+					},
+				},
+				"test:route2": {
+					Annotations: map[string]string{
+						"haproxy.router.openshift.io/timeout": "35s",
+					},
+				},
+				"test:route3": {
+					Annotations: map[string]string{
+						"haproxy.router.openshift.io/timeout-tunnel": "10m",
+					},
+				},
+				"test:route4": {},
+			},
+			annotation: "haproxy.router.openshift.io/timeout",
+			pattern:    `[1-9][0-9]*(us|ms|s|m|h|d)?`,
+			values:     []string{"30s"},
+			expected:   "5m",
+		},
+		{
+			name: "Default timeout is maximum",
+			state: map[ServiceAliasConfigKey]ServiceAliasConfig{
+				"test:route1": {},
+				"test:route2": {},
+				"test:route3": {},
+			},
+			annotation: "haproxy.router.openshift.io/timeout",
+			pattern:    `[1-9][0-9]*(us|ms|s|m|h|d)?`,
+			values:     []string{"30s"},
+			expected:   "30s",
+		},
```

Comment:
Should at least one of the routes specify a value?  It makes sense to me to have a test case for "some routes specify timeouts, but the default timeout is maximum" .

Reply:
By the way, what happens if you have 0 routes?

Reply:
> Should at least one of the routes specify a value? It makes sense to me to have a test case for "some routes specify timeouts, but the default timeout is maximum" .

The goal of maxTimeoutFirstMatchedAndClipped is to get the maximum timeout among the ones set on the lowest layer backends, rather than the maximum of all values provided to the function. For instance, if a route has a timeout annotation set, it will take precedence over the default timeout, even if the default timeout is greater. I updated the godoc for the function to highlight this.

> By the way, what happens if you have 0 routes?

Unit test case added to cover this case (`Empty state`).